src/main/java/com/aws/greengrass/disk/spool/DiskSpoolDAO.java

```diff
@@ -51,16 +53,36 @@ public class DiskSpoolDAO {
             RetryUtils.RetryConfig.builder().initialRetryInterval(Duration.ofSeconds(1L))
                     .maxRetryInterval(Duration.ofSeconds(3L)).maxAttempt(3)
                     .retryableExceptions(Collections.singletonList(SQLTransientException.class)).build();
+    private volatile Connection dbConnection;
 
+    private final ReentrantReadWriteLock dbConnectionLock = new ReentrantReadWriteLock();
     /**
      * This method will construct the database path.
      * @param paths The path to the working directory
      * @throws IOException when fails to set up the database
+     * @throws SQLException when fails to make database connection
      */
+
     @Inject
-    public DiskSpoolDAO(NucleusPaths paths) throws IOException {
+    public DiskSpoolDAO(NucleusPaths paths) throws IOException, SQLException {
         databasePath = paths.workPath(PERSISTENCE_SERVICE_NAME).resolve(DATABASE_FILE_NAME);
         url = String.format(DATABASE_CONNECTION_URL, databasePath);
+        init();
+    }
+
+    /** Initialize the singleton database connection.
+     *
+     * @throws SQLException when fails to make database connection
+     */
+    synchronized void init() throws SQLException {
+        try (LockScope ignored = LockScope.lock(dbConnectionLock.writeLock())) {
+            if (dbConnection == null || dbConnection.isClosed() || !isConnectionHealthy(dbConnection)) {
+                if (dbConnection != null && !dbConnection.isClosed()) {
+                    dbConnection.close(); // close the existing connection if it's unhealthy
```

Comment:
this can throw, i think we'd want to log the error and just go with a new connection